### PR TITLE
GT-1423 Fix a dispatcher deadlock caused by using runBlocking

### PR DESCRIPTION
**Pull request by frett**
This PR is to fix a deadlock caused by `runBlocking` deadlocking the dispatcher by waiting for a different lock to be released.

- `runBlocking` locks the current thread to the `runBlocking` call and doesn't allow the thread to be used by other coroutines.
- executing `mutex.read.unlock()` on a coroutines dispatcher thread will block that thread and prevent it from being used by other coroutines until `unlock()` finishes
- if `unlock()` has to wait for the `stateMutex` to be released by another coroutine it could deadlock if there are no remaining dispatcher threads to process other coroutines

This could all combine to a deadlock scenario if there is high volume locking/unlocking happening on a `ReadWriteMutex`

### CODE REVIEW

File: gto-support-kotlin-coroutines/src/test/kotlin/org/ccci/gto/android/common/kotlin/coroutines/ReadWriteMutexTest.kt [hunk 1]
```diff
@@ -139,6 +141,31 @@ class ReadWriteMutexTest {
         mutex.read.unlock()
     }
 
+    @Test(timeout = 10000)
+    fun `GT-1423 readUnlock causes deadlock from runBlocking usage`() = runBlocking {
+        mutex.write.lock()
+
+        launch(Dispatchers.Unconfined) {
+            expect(1)
```
Comment (wrandall22) [inline gto-support-kotlin-coroutines/src/test/kotlin/org/ccci/gto/android/common/kotlin/coroutines/ReadWriteMutexTest.kt:149]: Do these `expect` calls refer to the next line? e.g. Expect `mutex.read.lock()` to be called once?

Comment (frett) [reply gto-support-kotlin-coroutines/src/test/kotlin/org/ccci/gto/android/common/kotlin/coroutines/ReadWriteMutexTest.kt:149]: they count through the expected order of execution. the `expect()` and `finish()` methods can be seen at the bottom of the test.

I just updated the description describing what I'm trying to fix in case you hadn't seen it

**Comment by frett** [comment]
I'm looking into the failed test right now, I think it might just be a flaky test due to an edge condition. I'm going to see if I can adjust it a bit to better test the issue.

**Comment by wrandall22** [comment]
Looks like the last commit breaks a test. The stacktrace isn't very helpful for determining which test, but I'd guess it's the one you wrote to show the deadlock occurring.

**Comment by wrandall22** [comment]
Seems like the new test should have changed (or at least the description) if the last commit should fix the deadlock, right?

**Comment by frett** [comment]
I package up the test results as a zip, so I was able to download them. it was actually a different test that broke due to a race condition in the test itself 🤦

**Review by wrandall22 — APPROVED** [review]
